agreementbot/secrets/plugin_errors.go

```diff
@@ -0,0 +1,174 @@
+	ResponseCode int
+	VaultResponse string
+
+	// implement error interface
+	Err error
+}
+
+func (e *VaultError) Error() string {
+	return e.Error()
+}
+
+// ----- HELPER FUNCTION -----
+
+// takes one of the errors below and wraps it into a VaultError
+// that records the response code and vault response in addition to the error itself
+func WrapVaultError(err error) *VaultError {
+
+	// if no error, ignore 
+	if err == nil {
+		return nil
+	}
+
+	// case on the type of error
+	switch e := err.(type) {
+		case *PermissionDenied:
+			return &VaultError{ResponseCode: 403, VaultResponse: vaultRespToString(e.VaultResponse), Err: err}
+		case *Unauthenticated:
+			return &VaultError{ResponseCode: 401, VaultResponse: e.LoginError.Error(), Err: err}
+		case *VaultUnavailable:
+			return &VaultError{ResponseCode: 503, VaultResponse: e.VaultError.Error(), Err: err}
+		case *InvalidVaultResponse:
+			// find the vault response in the struct
+			var errString string
+			if e.ReadError != nil {
+				errString = e.ReadError.Error()
+			} else {
+				errString = string(e.VaultResponse)
+			}
+
+			return &VaultError{ResponseCode: 500, VaultResponse: errString, Err: err}
+		case *BadRequest:
+			return &VaultError{ResponseCode: e.ResponseCode, VaultResponse: vaultRespToString(e.VaultResponse), Err: err}
+		case *NoSecretFound: 
+			return &VaultError{ResponseCode: 404, VaultResponse: vaultRespToString(e.VaultResponse), Err: err}
+		case *Unknown: 
+			return &VaultError{ResponseCode: e.ResponseCode, VaultResponse: vaultRespToString(e.VaultResponse), Err: err}
+		default:
+			return nil
+	}
+}
+
+// ----- ERRORS -----
+
+// PermissionDenied - 403
+// the user provided correct credentials but does not have permission to access the resource
+type PermissionDenied struct {
+	VaultResponse map[string][]string
+	HttpMethod    string
+	SecretPath    string
+}
+
+func (e *PermissionDenied) Error() string {
+	return fmt.Sprintf("Permission denied, user does not have %s access to %s.", e.HttpMethod, e.SecretPath)
+}
+
+// Unauthenticated - 404
+// the user provided wrong credentials, was not able to authenticate with the exchange
+type Unauthenticated struct {
+	LoginError   error
+	ExchangeUser string
+}
+
+func (e *Unauthenticated) Error() string {
+	return fmt.Sprintf("Unable to authenticate user \"%s\" with exchange: \"%s\".", e.ExchangeUser, e.LoginError.Error())
+}
+
+// VaultUnavailable - 503
+// the vault service is unavailable
+type VaultUnavailable struct {
+	VaultError error
+}
+
+func (e *VaultUnavailable) Error() string {
+	return fmt.Sprintf("Secrets manager service unavailable: \"%s\"", e.VaultError.Error())
+}
+
+// InvalidVaultResponse - 500
+// the plugin was unable to read and/or parse the vault's response
+type InvalidVaultResponse struct {
+	ReadError     error
+	VaultResponse []byte
+	HttpMethod    string
+	SecretPath    string
+}
+
+func (e *InvalidVaultResponse) Error() string {
+	if e.ReadError != nil {
```

Comment:
This is a little odd, the absence of ReadError means parse error. We should be more explicit. Maybe add a ParseError field to the struct? ...not sure what is the best way.